src/Compilers/Core/Portable/Serialization/SerializationThreadPool.cs

```diff
@@ -0,0 +1,170 @@
+        /// a new thread is created. This is meant as a stop-gap measure for workloads that would otherwise be
+        /// creating a new thread for every work item.
+        /// </summary>
+        /// <remarks>
+        /// This class is derived from <see href="https://github.com/dotnet/machinelearning/blob/ebc431f531436c45097c88757dfd14fe0c1381b3/src/Microsoft.ML.Core/Utilities/ThreadUtils.cs">dotnet/machinelearning</see>.
+        /// </remarks>
+        private static class ImmediateBackgroundThreadPool
+        {
+            /// <summary>How long should threads wait around for additional work items before retiring themselves.</summary>
+            private static readonly TimeSpan s_idleTimeout = TimeSpan.FromSeconds(1);
+
+            /// <summary>The queue of work items. Also used as a lock to protect all relevant state.</summary>
+            private static readonly Queue<(Delegate, object?, TaskCompletionSource<object?>)> s_queue = new();
+
+            /// <summary>The number of threads currently waiting for work to arrive.</summary>
+            private static int s_availableThreads = 0;
+
+            /// <summary>
+            /// Queues an <see cref="Action"/> delegate to be executed immediately on another thread,
+            /// and returns a <see cref="Task"/> that represents its eventual completion. The task will
+            /// always end in the <see cref="TaskStatus.RanToCompletion"/> state; if the delegate throws
+            /// an exception, it'll be allowed to propagate on the thread, crashing the process.
+            /// </summary>
+            public static Task<object?> QueueAsync(Func<object?> threadStart) => QueueAsync((Delegate)threadStart, null);
+
+            /// <summary>
+            /// Queues an <see cref="Action{Object}"/> delegate and associated state to be executed immediately on another thread,
+            /// and returns a <see cref="Task"/> that represents its eventual completion. The task will
+            /// always end in the <see cref="TaskStatus.RanToCompletion"/> state; if the delegate throws
+            /// an exception, it'll be allowed to propagate on the thread, crashing the process.
+            /// </summary>
+            public static Task<object?> QueueAsync(Func<object?, object?> threadStart, object? state) => QueueAsync((Delegate)threadStart, state);
+
+            private static Task<object?> QueueAsync(Delegate threadStart, object? state)
+            {
+                // Create the TaskCompletionSource used to represent this work.
+                // Call sites only care about completion, not about the distinction between
+                // success and failure and do not expect exceptions to be propagated in this manner,
+                // so only SetResult is used.
+                var tcs = new TaskCompletionSource<object?>(TaskCreationOptions.RunContinuationsAsynchronously);
```

Comment:
doc why we want RunContinuationsAsynchronously here.

Reply:
➡️ Added this documentation, and removed the incorrect documentation about only using `SetResult`